File: js/src/base64url.ts
/** Base64url (RFC 4648 §5) encode/decode for Uint8Array, dependency-free, browser + node. */

import { SpectrlDecodeError } from "./errors.js";

const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

const LOOKUP = (() => {
  const t = new Int16Array(128).fill(-1);
  for (let i = 0; i < ALPHABET.length; i++) t[ALPHABET.charCodeAt(i)] = i;
  return t;
})();

/** Encode bytes to base64url without padding. */
export function b64urlEncode(data: Uint8Array): string {
  let out = "";
  let i = 0;
  for (; i + 3 <= data.length; i += 3) {
    const n = (data[i]! << 16) | (data[i + 1]! << 8) | data[i + 2]!;
    out += ALPHABET[(n >>> 18) & 63]! + ALPHABET[(n >>> 12) & 63]! + ALPHABET[(n >>> 6) & 63]! + ALPHABET[n & 63]!;
  }
  const rem = data.length - i;
  if (rem === 1) {
    const n = data[i]! << 16;
    out += ALPHABET[(n >>> 18) & 63]! + ALPHABET[(n >>> 12) & 63]!;
  } else if (rem === 2) {
    const n = (data[i]! << 16) | (data[i + 1]! << 8);
    out += ALPHABET[(n >>> 18) & 63]! + ALPHABET[(n >>> 12) & 63]! + ALPHABET[(n >>> 6) & 63]!;
  }
  return out;
}

/** Bits of the final character that fall outside the decoded bytes. A canonical
 * encoding leaves them zero, so exactly one spelling exists per payload. */
const TAIL_MASK: Record<number, number> = { 2: 0b001111, 3: 0b000011 };

/** Decode a canonical unpadded base64url string. Strict: rejects characters
 * outside the base64url alphabet, `=` padding, impossible lengths, and
 * non-zero unused trailing bits, matching the Python reference
 * implementation. Each payload therefore has exactly one valid spelling. */
export function b64urlDecode(s: string): Uint8Array {
  if (s.includes("=")) throw new SpectrlDecodeError("invalid base64url payload: must be unpadded");
  const len = s.length;
  if (len % 4 === 1) throw new SpectrlDecodeError("invalid base64url payload: impossible length");
  const outLen = Math.floor((len * 6) / 8);
  const out = new Uint8Array(outLen);
  let bits = 0;
  let acc = 0;
  let oi = 0;
  for (let i = 0; i < len; i++) {
    const v = LOOKUP[s.charCodeAt(i)] ?? -1;
    if (v < 0) throw new SpectrlDecodeError(`invalid base64url character: ${JSON.stringify(s[i])}`);
    acc = (acc << 6) | v;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      out[oi++] = (acc >>> bits) & 0xff;
    }
  }
  const mask = TAIL_MASK[len % 4];
  if (mask !== undefined && (LOOKUP[s.charCodeAt(len - 1)]! & mask) !== 0) {
    throw new SpectrlDecodeError("invalid base64url payload: non-zero trailing bits");
  }
  return out;
}

File: rust/src/base64url.rs
//! Canonical unpadded RFC 4648 base64url.

use crate::error::{Result, bail};

const ALPHABET: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

pub fn encode(data: &[u8]) -> String {
    let mut out = String::with_capacity(data.len().div_ceil(3) * 4);
    for chunk in data.chunks(3) {
        let b = [
            chunk[0],
            *chunk.get(1).unwrap_or(&0),
            *chunk.get(2).unwrap_or(&0),
        ];
        let n = (b[0] as u32) << 16 | (b[1] as u32) << 8 | b[2] as u32;
        let chars = chunk.len() + 1;
        for i in 0..chars {
            out.push(ALPHABET[(n >> (18 - 6 * i) & 63) as usize] as char);
        }
    }
    out
}

fn sextet(c: u8) -> Option<u32> {
    Some(match c {
        b'A'..=b'Z' => c - b'A',
        b'a'..=b'z' => c - b'a' + 26,
        b'0'..=b'9' => c - b'0' + 52,
        b'-' => 62,
        b'_' => 63,
        _ => return None,
    } as u32)
}

/// Decode, rejecting padding, foreign characters, impossible lengths and a
/// final character with nonzero unused bits.
pub fn decode(text: &str) -> Result<Vec<u8>> {
    let bytes = text.as_bytes();
    if bytes.contains(&b'=') {
        bail!("invalid base64url payload: padding is not allowed");
    }
    if bytes.len() % 4 == 1 {
        bail!("invalid base64url payload: impossible length");
    }
    let mut out = Vec::with_capacity(bytes.len() / 4 * 3 + 2);
    for chunk in bytes.chunks(4) {
        let mut n = 0u32;
        for (i, &c) in chunk.iter().enumerate() {
            let Some(v) = sextet(c) else {
                bail!("invalid base64url payload: unexpected character");
            };
            n |= v << (18 - 6 * i);
        }
        let produced = chunk.len() - 1;
        let unused_mask = match chunk.len() {
            2 => 0xffff,
            3 => 0xff,
            _ => 0,
        };
        if n & unused_mask != 0 {
            bail!("invalid base64url payload: noncanonical trailing bits");
        }
        for i in 0..produced {
            out.push((n >> (16 - 8 * i)) as u8);
        }
    }
    Ok(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trip_lengths() {
        for len in 0..20usize {
            let data: Vec<u8> = (0..len as u8)
                .map(|b| b.wrapping_mul(37).wrapping_add(250))
                .collect();
            assert_eq!(decode(&encode(&data)).unwrap(), data);
        }
        assert!(decode("oQAB").is_ok());
        assert!(decode("oR").is_err());
        assert!(decode("oQ").is_ok());
        assert!(decode("A").is_err());
        assert!(decode("oQ==").is_err());
    }
}
